### ISSUE TITLE
DeepSpeed learning rate scheduler not working

### ISSUE THREAD

**Issue by gahdritz** [PyTorchLightning/pytorch-lightning#11694]
## 🐛 Bug

PyTorch Lightning does not appear to be using a learning rate scheduler specified in the DeepSpeed config as intended. It increments the learning rate only at the end of each epoch, rather than once per step. Additionally, it uses the wrong learning rate for the very first epoch.

### To Reproduce

I used the following DeepSpeed config:

```
{
  "optimizer": {
    "type": "Adam",
    "params": {
      "lr": 0.001,
      "eps": 1e-05
    }
  },
  "scheduler": {
      "type": "WarmupLR",
      "params": {
        "warmup_min_lr": 0,
        "warmup_max_lr": 0.001,
        "warmup_num_steps": 3000,
        "warmup_type": "linear"
      }
  },
  "fp16": {
    "enabled": false,
    "min_loss_scale": 1
  },
  "amp": {
    "enabled": false,
    "opt_level": "O2"
  },
  "bfloat16": {
    "enabled": true
  },
  "zero_optimization": {
    "stage": 2,
    "cpu_offload": false,
    "contiguous_gradients": false
  },
  "activation_checkpointing": {
    "partition_activations": true,
    "cpu_checkpointing": false,
    "profile": false
  },
  "gradient_clipping": 0.1
}
```

I then used a `pytorch_lightning.callbacks.lr_monitor.LearningRateMonitor` to monitor the learning rate. This is the result:

![image](https://user-images.githubusercontent.com/26167235/152055476-9e63ff63-4d73-47ed-abc8-f42745ff7f40.png)

For the first epoch, the learning rate is 0.001. For the next one, it's zero. Thereafter, the learning rate is warmed up just once per epoch, rather than once per step.

It seems to me like the DeepSpeed LR is being updated on the "epoch" interval rather than the "step" one.

### Environment

* CUDA:
        - GPU:
        - available:         False
        - version:           11.3
* Packages:
        - numpy:             1.21.1
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.1+cu113
        - pytorch-lightning: 1.5.9
        - tqdm:              4.62.2
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                -
        - processor:         x86_64
        - python:            3.7.0
        - version:           #1 SMP Wed Dec 1 05:10:33 EST 2021

cc @SeanNaren @awaelchli @rohitgr7

**Comment by rohitgr7** [comment]
well, to decide at what level lr schedulers should operate, you need to set interval inside scheduler_config when using `configure_optimizer`, but within deepspeed config, it doesn't seems to be possible. Can you try using the [configure_optimizer](https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_module.html#configure-optimizers) method instead?

I think if it's defined within the config, the default interval is set to epoch, so it will update after every epoch.
https://github.com/PyTorchLightning/pytorch-lightning/blob/86b177ebe5427725b35fde1a8808a7b59b8a277a/pytorch_lightning/strategies/deepspeed.py#L483-L491

**Comment by gahdritz** [comment]
I'm actually not getting that warning---it looks like it's been added since 1.5.9. You're right that the issue is easy to work around, but isn't it still a bug? The Lightning documentation claims to support DeepSpeed optimizers/learning rate schedulers, but they behave differently in Lightning than expected.

**Comment by rohitgr7** [comment]
yes, but as a framework, since the schedulers are controlled internally, and deepspeed just takes in a scheduler input, it can't really tell when it should update. It does support deepspeed out of the box but for some cases, we need to make adjustments at some level to make it work.

cc @SeanNaren

**Comment by SeanNaren** [comment]
@gahdritz I agree that this is a bug. @rohitgr7 can we make an assumption that if the user passes the deepspeed config that we automatically set this interval to step? I think all schedulers for DeepSpeed are called every step rather than every epoch.

I'm unfamiliar with the code format of the scheduling logic now to be the person to make this change!

**Comment by SeanNaren** [comment]
@gahdritz once #11788 is merged this should be fixed!

**Comment by antoiloui** [comment]
Hello, I had the exact same issue as @gahdritz. 

I've installed the latest master branch but the problem doesn't seem fixed. The learning rate is now constant from the start and the warmup seems to be ignored (see screenshot below).
<img width="847" alt="Capture d’écran 2022-02-15 à 12 41 40" src="https://user-images.githubusercontent.com/32681432/154055388-ceb7eab1-ce09-4364-b8f4-dc8cf39ee88c.png">

I'm using the following deepspeed config:
```json
{
    "zero_optimization": {
       "stage": 2,
       "offload_optimizer": {
           "device": "cpu",
           "pin_memory": true
       },
       "allgather_partitions": true,
       "allgather_bucket_size": 2e8,
       "reduce_scatter": true,
       "reduce_bucket_size": 2e8,
       "overlap_comm": true,
       "contiguous_gradients": true
    },
    "fp16": {
        "enabled": true,
        "loss_scale": 0,
        "loss_scale_window": 1000,
        "initial_scale_power": 16,
        "hysteresis": 2,
        "min_loss_scale": 1
    },
    "optimizer": {
        "type": "AdamW",
        "params": {
            "lr": 2e-5,
            "betas": [0.9, 0.999],
            "eps": 1e-7,
            "weight_decay": 3e-7
        }
    },
    "scheduler": {
        "type": "WarmupLR",
        "params": {
            "warmup_min_lr": 0,
            "warmup_max_lr": 2e-5,
            "warmup_num_steps": 100,
            "warmup_type": "linear"
        }
    },
    "steps_per_print": 2000,
    "wall_clock_breakdown": false
}
```

I'm also experiencing the same issue with the "WarmupDecayLR": the learning rate stays constant through the whole run.

Any idea of what is happening?

**Comment by SeanNaren** [comment]
@awaelchli this is the recommended way yes, however many users just rely on the config file as per deepspeed's recommendation.

@antoiloui seems like the bug still exists. It would really help to get together a bug report model example that uses this config: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pl_examples/bug_report/bug_report_model.py

If I do get time I'll try to debug this myself!

**Comment by rohitgr7** [comment]
hey @antoiloui !

can you try out #12031 patch?

**Comment by antoiloui** [comment]
@awaelchli indeed, the only way I was able to make a scheduler work was by defining it in `configure_optimizers`. 

I couldn't make the default deepspeed schedulers work though (e.g., WarmupLR) but the hf ones seem to work fine with DeepSpeedCPUAdam. Here is my current `configure_optimizers` function:
```python
from pytorch_lightning.plugins import DeepSpeedPlugin
from deepspeed.ops.adam import DeepSpeedCPUAdam
from transformers.optimization import AdamW, get_scheduler

def configure_optimizers(self):
        # Create groups of parameters to optimize (with weight decay on all parameters other than bias and layer normalization terms).
        no_decay = ['bias', 'LayerNorm.bias', 'LayerNorm.weight']
        optimizer_grouped_params = [
            {'params': [p for n, p in self.named_parameters() if not any(nd in n for nd in no_decay)], 'weight_decay': self.hparams.weight_decay},
            {'params': [p for n, p in self.named_parameters() if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
        ]
        # Configure optimizer.
        if isinstance(self.trainer.training_type_plugin, DeepSpeedPlugin):
            optimizer = DeepSpeedCPUAdam(optimizer_grouped_params, adamw_mode=True, lr=self.hparams.lr, betas=(0.9, 0.999), eps=1e-7)
        else:
            optimizer = AdamW(optimizer_grouped_params, lr=self.hparams.lr, betas=(0.9, 0.999), eps=1e-7)
        # Configure learning rate scheduler.
        scheduler = get_scheduler(name=self.hparams.scheduler_type, optimizer=optimizer, num_warmup_steps=self.warmup_steps, num_training_steps=self.total_steps)
        scheduler = {"scheduler": scheduler, "interval": "step", "frequency": 1}
        return [optimizer], [scheduler]
```

@SeanNaren @rohitgr7 I'm in a very busy week, will try it as soon as I have some time.

**Comment by rohitgr7** [comment]
with configure_optimizers it will work as expected, but for deepspeed, it's sort of like an explicit behavior where the scheduler provided within the config should work on step level, and to provide full support for it, Lightning should handle it.

Waiting for your confirmation :)

**Comment by luohao123** [comment]
How to using cosine lr scheduler?
